homeassistant/components/hassmpris/media_player.py

```diff
@@ -0,0 +1,674 @@
+        _LOGGER.debug("%s: new properties: %s", self.name, props)
+
+        feats = self._attr_supported_features
+        if props.HasField("CanControl"):
+            if not props.CanControl:
+                feats = 0
+            else:
+                feats = SUPPORTED_MINIMAL
+
+        update_state = False
+
+        for name, bitwisefield in {
+            "CanPlay": MediaPlayerEntityFeature.PLAY,
+            "CanPause": MediaPlayerEntityFeature.PAUSE,
+            "CanSeek": MediaPlayerEntityFeature.SEEK,
+            "CanGoNext": MediaPlayerEntityFeature.NEXT_TRACK,
+            "CanGoPrevious": MediaPlayerEntityFeature.PREVIOUS_TRACK,
+        }.items():
+            if props.HasField(name):
+                val = getattr(props, name)
+                if val:
+                    feats = feats | bitwisefield
+                else:
+                    feats = feats & ~bitwisefield
+
+        if feats != self._attr_supported_features:
+            _LOGGER.debug(
+                "%s: new feature bitfield: (%s) %s",
+                self.name,
+                feats,
+                _feat2bitfield(feats, MediaPlayerEntityFeature),
+            )
+            self._attr_supported_features = feats
+            update_state = True
+
+        if props.HasField("Rate") and props.Rate != self._attr_playback_rate:
+            _LOGGER.debug("%s: new rate: %s", self.name, props.Rate)
+            self._attr_playback_rate = props.Rate
+            update_state = True
+
+        if update_state and self.hass:
+            await self.async_update_ha_state(True)
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Return entity specific state attributes."""
+        if self.state == STATE_OFF:
+            return {}
+        return {ATTR_PLAYBACK_RATE: self._attr_playback_rate}
+
+
+class EntityManager:
+    """
+    The entity manager manages MPRIS media player entities.
+
+    This class is responsible for maintaining the known player entities
+    in sync with the state as reported by the server, as well as keeping
+    tabs of newly-appeared players and players that have gone.
+    """
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+        mpris_client: hassmpris_client.AsyncMPRISClient,
+        async_add_entities: AddEntitiesCallback,
+    ) -> None:
+        """
+        Initialize the entity manager.
+
+        Parameters:
+          hass: the HomeAssistant singleton
+          config_entry: the configuration entry associated with
+                        this component (or integration?)
+          async_add_entities: callback to add entities async
+        """
+        self.hass = hass
+        self.config_entry = config_entry
+        self.async_add_entities = async_add_entities
+        self._client = mpris_client
+        self._players: dict[str, HASSMPRISEntity] = {}
```

Comment:
We generally don't allow storing entities in a shared container

Reply:
It's not a shared container — this container is strictly per-instance.  There is exactly and always _one_ instance of `EntityManager` per HASS MPRIS integration entry.

However, I am interested in knowing if you know of an alternative.  Lord knows I wish I didn't have to keep the state between the config entries and the `EntityManager` in sync!

Reply:
I'm assuming this comment is stale and resolving it.  That said, I would love to know how / if I should change this — the purpose of this object is to be able to look up the entities based on ID, and if there is another way to do this without this variable, it would be great to know.  Thanks in advance.

Reply:
We no longer store references to entities in this class -- which has now become a DataUpdateCoordinator!